### ISSUE TITLE
Unnecessary memory allocations for `hist` method?

### ISSUE THREAD

**Issue by thvasilo** [dmlc/xgboost#4915]
The hist method makes heavy use of a [HistCollection](https://github.com/dmlc/xgboost/blob/f0064c07ab70f1c770bb0e9eb64fe85afb3a3d73/src/common/hist_util.h#L440) container that is supposed to maintain for each node (leaf) id a collection of histograms.

The backing data structure is [a vector of vectors of `GradStatHist`](https://github.com/dmlc/xgboost/blob/f0064c07ab70f1c770bb0e9eb64fe85afb3a3d73/src/common/hist_util.h#L476) objects.

Now, whenever we add a new node id to this collection, we resize this vector to hold enough elements so that _the highest nid provide_ is a valid index.

From some print debugging, printing out nodeid [here](https://github.com/dmlc/xgboost/blob/f0064c07ab70f1c770bb0e9eb64fe85afb3a3d73/src/tree/updater_quantile_hist.cc#L615) (printing the node id of every node being synced), it seems like node ids _are non-contiguous_. For example they might be [1,3,5,7,9,11,13]. However whenever we expand the size of the container, [it looks like](https://github.com/dmlc/xgboost/blob/f0064c07ab70f1c770bb0e9eb64fe85afb3a3d73/src/common/hist_util.h#L462) we allocate memory for _all_ node ids, regardless of whether they represent an actual node or not. Is this the case, or are node ids actually contiguous, thereby it's necessary to always pre-populate every element?

As far as I understand the 2D data vector has dimensions `<MAX_NODE_ID> x <NBINS>`, where [uint32_t nbins = gmat.cut.Ptrs().back();](https://github.com/dmlc/xgboost/blob/f0064c07ab70f1c770bb0e9eb64fe85afb3a3d73/src/tree/updater_quantile_hist.cc#L536). For example for the `rcv1` dataset that has ~42k features creates 748,378 bins per node, which each correspond I think to two float values per bin. So every time we add a redundant item in the `_hist` vector we are wasting num_bins * 2 float values, which, when we have say millions of features can be very significant.

So my impression is that we are currently using twice as much memory as is necessary, because we initialize every vector in the 2D vector of histogram values with `nbins` values, regardless of the fact that half of the node ids are not used (?).

Is my analysis correct? AFAIK in approx we get around this issue but having a translation between nodeid to a "working_id" or index, that is contiguous. That allows us to pack all histogram values in a contiguous array of doubles, of total size <num_features*num_bins*2> (one for gradients, one for hessians), and then use indexing tricks to get the right values.

I'll note that in my distributed experiments the hist method produces OOM errors for datasets that are easily handled by the approx method, and I think this might be one of the reasons.

**Comment by thvasilo** [comment]
I noticed this as part of looking into #4679. While the second per-node sync seems straightforward to fix (see [this WIP](https://github.com/thvasilo/xgboost/commit/8196a16e5732b579addb5e5b7ac3339bd0545758#diff-10aac201acca00858b390ff0d679434fR655)) getting rid of the first loop could entail changes to how we build and use `hist_` which would requires a lot of effort to untangle.

**Comment by trivialfis** [comment]
I haven't look into your references in detail yet.  I believe the reason of nodes not being continuous is the use of `pruner`.  On GPU side, we don't use pruner, instead we just don't apply the split when it's gain is not sufficient.  See  #4874.  But will look into your investigation later, thanks for the detailed explanation.

**Comment by thvasilo** [comment]
Thanks @trivialfis . It's probably the case that something useful is done to every element in the 2D vector, because yesterday I tried to only populate the new node id and not all ids up to the new (<= id) and that led to crashes. 

Weirdly enough the single machine code crashed, but I think I was able to run the training distributed.

I'd say the reason for the excessive memory use of `hist` is _probably_ not what I described, but there's definitely something inefficient going on, especially for high dimensional data.

**Comment by hcho3** [comment]
@thvasilo The generation of quantized matrix (`gmat`) is another source of inefficiency. Basically we use quantile points to convert matrix elements in the input matrix into bin indices. To save memory, we'll need to perform conversion in out-of-memory fashion:

1. Load data
2. Generate quantile sketches
3. Now load data again, this time performing quantization on the fly.

**Comment by RAMitchell** [comment]
I would like to tackle memory usage issues for GPU and CPU hist algorithms with a common strategy. I think this will go hand in hand with further re-factoring of DMatrix.

**Comment by hcho3** [comment]
@RAMitchell Good to know. Anything I can do on my end to make the job easier?

**Comment by RAMitchell** [comment]
I probably can't work on this for a while. I would start by removing extra layers inside the DMatrix pipeline, it is extremely difficult to reason about and work with. I think the DataSource classes are basically redundant and their functionality should live inside the respective DMatrix classes. I would also like to see a lot of the code constructing a DMatrix live inside actual constructors instead of manually constructing DataSource objects.

I am interested in trying to create a common iterator layer (without memory allocation) over input data sources (e.g. csv, numpy 2d, libsvm). If this is possible we could calculate quantiles directly on external data, or even train directly on external data with DMatrix as a thin wrapper.

**Comment by hcho3** [comment]
@RAMitchell That sounds awesome. Yes, DMatrix class can be improved a lot.

**Comment by hcho3** [comment]
> The backing data structure is a vector of vectors of GradStatHist objects.
> Now, whenever we add a new node id to this collection, we resize this vector to hold enough elements so that the highest nid provide is a valid index.

This is fixed in #5156
